Add util/kernel package

brycekahle opened the pull request:
### What does this PR do?

Extracts Linux kernel inspection functions to a common package outside of `pkg/ebpf` and `pkg/process/util`. This functionality is not ebpf/process-agent specific.

- `Version` type abstracts the kernel version math in a clean way.
- `IsDebugFSMounted` properly returns an error
- Use `metadata/host` package to get platform rather than method that runs commands. This is consistent with the rest of the agent.
- Adds function `FindHeaderDirs` to find kernel header directories that correspond to the running version.

### Motivation

Cleaner code and logical placement of functionality. 

### Additional Notes


### Describe your test plan

- automated tests exist for version logic
- sysprobe fails to start if `debugfs` is not mounted.
- sysprobe fails to start on Ubuntu with kernel version between 4.4.114 and 4.4.127
- kernel headers are found on all systems where correct headers are installed or kheaders module is enabled.

brycekahle committed:
Add util/kernel package with Version type to abstract away version math

brycekahle committed:
Move IsDebugFSMounted to kernel package and return an error

brycekahle committed:
Use better internal method of obtaining the platform string

brycekahle committed:
Remove linux-only kernel version output from nettop

brycekahle committed:
Cache host kernel version

brycekahle committed:
Add kernel header search and version validation

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -90,8 +90,8 @@ const (
-	if mounted, msg := util.IsDebugfsMounted(); !mounted {
-		return nil, fmt.Errorf("%s: %s", "system-probe unsupported", msg)
+	if mounted, err := kernel.IsDebugFSMounted(); !mounted {
+		return nil, fmt.Errorf("%s: %s", "system-probe unsupported", err)
```
just to verify: will this continue to print out the instructions for mounting debugfs ?

brycekahle replied:
Correct.

brycekahle committed:
go mod tidy

leeavital commented on pkg/util/kernel/headers.go:
```diff
@@ -0,0 +1,157 @@
+		return nil, fmt.Errorf("unable to get host metadata")
+	}
+
+	dirs := []string{kernelModulesPath}
```
I was a bit thrown by how this was written, could we do something like:

```
dirs := []string{fmt.Sprintf(kernelModulesPath, hi.KernelVersion}}
if hi.Platform == "debian" {
   dirs = append(dirs, fmt.Sprintf("/lib/modules/%s/source", hi.KernelVersion))
}
```

brycekahle replied:
Sure.

leeavital commented on pkg/util/kernel/headers.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func parseHeaderVersion(r io.Reader) (Version, error) {
+	re := regexp.MustCompile(`^#define[\t ]+LINUX_VERSION_CODE[\t ]+(\d+)$`)
```
nit: this function gets called a lot, so this regexp should be pulled out to a const

brycekahle committed:
Run go fmt -s

hmahmood commented on pkg/network/dns_snooper_test.go:
```diff
@@ -3,52 +3,34 @@
 	dnsTimeout time.Duration,
 ) (*manager.Manager, *SocketFilterSnooper) {
-	currKernelVersion, err := ebpf.CurrentKernelVersion()
+	currKernelVersion, err := kernel.HostVersion()
```
Wouldn't `kernel.Version()` read better?

brycekahle replied:
`kernel.Version` is already a type.

leeavital commented on pkg/util/kernel/headers.go:
```diff
@@ -0,0 +1,157 @@
+
+	if os.IsNotExist(err) {
+		// as last resort, look for sysfs headers
+		if dir, syserr := getSysfsHeaderDirs(hv); syserr == nil {
```
Do we need to worry about removing the temp path this creates?

brycekahle replied:
We want it stick around, unless we want to extract the headers on every run on system-probe.

brycekahle replied:
I added code to remove it in the case that it fails to validate properly, so we don't attempt to use it again.

leeavital commented on pkg/util/kernel/headers.go:
```diff
@@ -0,0 +1,157 @@
+const sysfsHeadersPath = "/sys/kernel/kheaders.tar.xz"
+const kernelModulesPath = "/lib/modules/%s/build"
+
+func FindHeaderDirs() ([]string, error) {
```
It would be useful to explain in a comment which branches of this get triggered when -- e.g. when the kernelModules path has the headers, when the sysfsHeadersPath is populated, when modprobe gets used, and anything similar

brycekahle committed:
Add comment to FindHeaderDirs

brycekahle committed:
Use package var for regexp

brycekahle committed:
Format in-place rather that two loops

brycekahle committed:
Remove temp dir in error case, so we don't try it again

brycekahle committed:
Ignore error from unloadKHeadersModule

brycekahle committed:
Put header dir search test behind INTEGRATION flag

The deb/rpm test stages use non-package based headers for system-probe build, which do not match the host.

p-lambert commented on pkg/security/probe/probe_bpf.go:
```diff
@@ -63,7 +64,7 @@ type Probe struct {
 	syscallMonitor    *SyscallMonitor
 	loadController    *LoadController
-	kernelVersion     uint32
+	kernelVersion     kernel.Version
```
👍

sunhay commented on pkg/ebpf/utils_linux.go:
```diff
@@ -105,22 +100,20 @@ func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string)
-		}
+	if platform == "ubuntu" && kernelCode >= kernel.VersionCode(4, 4, 114) && kernelCode <= kernel.VersionCode(4, 4, 127) {
+		return false, fmt.Sprintf("Known bug for kernel %s on platform %s, see: \n- https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1763454", kernelCode, platform)
 	}
```
Its not clear to be why we don't need strings.Contains(`ubuntu`) anymore - could you clarify?

brycekahle replied:
The `platform` value that comes from the metadata package is more tightly controlled, so we can rely on the exact value `ubuntu`.

sunhay replied:
Thanks! 🙇

brycekahle committed:
Update DataDog/ebpf that makes kernel patch version optional

Fixes #6672

safchain commented:
We ran our kitchen functional tests, and they failed, we investigating why ... I ran them against master with success, I suspect a issue with latest patches of the ebpf lib

safchain commented:
We finally found the issue, it is on the runtime security side, we will fix it in a following PR.

brycekahle committed:
Update gopsutil to fix macos build
